kubeflow/tf-serving/tf-serving.libsonnet

```diff
@@ -1,152 +1,262 @@
-local k = import "k.libsonnet";
-local deployment = k.extensions.v1beta1.deployment;
-local container = deployment.mixin.spec.template.spec.containersType;
-local storageClass = k.storage.v1beta1.storageClass;
-local service = k.core.v1.service;
-local networkPolicy = k.extensions.v1beta1.networkPolicy;
-local networkSpec = networkPolicy.mixin.spec;
-
 {
+  // Parameters are intended to be late bound.
+  params:: {
+    name: null,
+    namespace: null,
+    numGpus: 0,
+    labels: {
+      app: $.params.name,
+    },
+    modelName: $.params.name,
+    modelPath: null,
+    // TODO(jlewi): We should probably default to a string.
+    // We might also want to have a separate boolean to indicate whether or not to use the httpProxy.
+    httpProxyImage: 0,
+
+    serviceType: "ClusterIP",
+
+    // If users want to override the image then can override defaultCpuImage and/or defaultGpuImage
+    // in which case the image used will still depend on whether GPUs are used or not.
+    // Users can also override modelServerImage in which case the user supplied value will always be used
+    // regardless of numGpus.
+    defaultCpuImage: "gcr.io/kubeflow-images-staging/tf-model-server:v20180227-master",
+    defaultGpuImage: "gcr.io/kubeflow-images-staging/tf-model-server-gpu:v20180305-pr362-7f250ae-5cc7",
+    modelServerImage: if $.params.numGpus == 0 then
+      $.params.defaultCpuImage
+    else
+      $.params.defaultGpuImage,
+
+
+    // Which cloud to configure the model for.
+    cloud:: null,
+  },
+
+  // Parameters that control S3 access
+  // params overrides s3params because params can be overwritten by the user to override the defaults.
+  s3params:: {
+    //  Name of the k8s secrets containing S3 credentials
+    s3SecretName: "",
+    // Name of the key in the k8s secret containing AWS_ACCESS_KEY_ID.
+    s3SecretAccesskeyidKeyName: "",
+
+    // Name of the key in the k8s secret containing AWS_SECRET_ACCESS_KEY.
+    s3SecretSecretaccesskeyKeyName: "",
+
+    // S3 region
+    s3AwsRegion: "us-west-1",
+
+    // TODO(jlewi): We should use util.toBool to automatically conver to actual boolean values.
+    // The use of strings is left over from when they were prototype parameters which only supports string type.
+
+    // true Whether or not to use https for S3 connections
+    s3UseHttps: "true",
+
+    // Whether or not to verify https certificates for S3 connections
+    s3VerifySsl: "true",
+
+    // URL for your s3-compatible endpoint.
+    s3Endpoint: "http://s3.us-west-1.amazonaws.com,",
+  } + $.params,
+
+
+  components:: {
+
+    all::
+      if $.params.cloud == "aws" then
```

Comment:
This would be better as params.s3Enable. The S3 functionality isn't specific to AWS.